File: concentration.pl
use strict;
use warnings;

package Concentration;

sub new {
    my ($class, $n, $t) = @_;
    my $self = {
        N            => $n,
        T            => $t,
        acquiredCards => [(0) x $n],
        P            => 0,
    };
    bless $self, $class;
    return $self;
}

sub concentration {
    my ($self, $a, $b, $c, $d) = @_;
    if ($self->{T}->[$a]->[$b] eq $self->{T}->[$c]->[$d]) {
        $self->{acquiredCards}->[$self->{P}] += 2;
    } else {
        $self->{P} = ($self->{P} + 1) % $self->{N};
    }
}

package main;

my ($H, $W, $N) = split ' ', <>;
chomp($H, $W, $N);
my @T;
for my $i (0 .. $H - 1) {
    my $line = <>;
    chomp($line);
    push @T, [split ' ', $line];
}
my $concentration = Concentration->new($N, \@T);
my $L = <>;
chomp($L);
while ($L-- > 0) {
    my ($a, $b, $A, $B) = split ' ', <>;
    $concentration->concentration($a - 1, $b - 1, $A - 1, $B - 1);
}
for my $i (0 .. $N - 1) {
    print $concentration->{acquiredCards}->[$i], "\n";
}

File: concentration.js
class Concentration {
    N = 0;
    T = null;
    acquiredCards = null;
    P = 0;
    getAcquiredCards(i) {
        return this.acquiredCards[i];
    }
    constructor(n, t) {
        this.N = n;
        this.T = t;;
        this.acquiredCards = new Array(n).fill(0);
        this.P = 0;
    }
    concentration(a, b, c, d) {
        if (T[a][b] === T[c][d]) this.acquiredCards[P] += 2;
        else P = -~P % this.N;
    }
}

const lines = require("fs").readFileSync("/dev/stdin", "utf8").trim().split("\n");
let _ = 0;
const [H, W, N] = lines[_++].split(' ').map(Number);
const T = [];
for (var i = 0; i < H; i++)T.push(lines[_++].split(' '));
const concentration = new Concentration(N, T);
let L = Number(lines[_++]);
while (L--) {
    let [a, b, A, B] = lines[row++].split(" ").map(x => Number(x) - 1);
    concentration.concentration(a, b, A, B);
}
for (var i = 0; i < N; i++)console.log(concentration.getAcquiredCards(i));
